plugins/source-terasology-modules/gatsby-node.js

```diff
@@ -92,78 +105,118 @@ exports.sourceNodes = async (
     await cache.set(lastFetchedKey, now.toISO());
+
+    fs.writeFileSync(
+      `${__dirname}/data.json`,
+      JSON.stringify(repositories, null, 2)
+    );
   }
 
   reporter.success(`[${PLUGIN_NAME}] Loaded ${repositories.length} modules.`);
 
-  let created = 0;
-  repositories
-    .forEach((repo) => {
-      if (!repo.moduleTxt) {
-        // skip non-module repositories
-        return;
-      }
+  const nodes = repositories.flatMap(async (repo) => {
+    if (!repo.moduleTxt) {
+      // skip non-module repositories
+      return [];
+    }
 
-      let moduleTxt;
+    let moduleTxt;
+    try {
+      moduleTxt = JSON.parse(repo.moduleTxt?.text);
+    } catch (err) {
+      console.warn(
+        `[${PLUGIN_NAME}] Could not parse 'module.txt' of ${repo.url}.`
+      );
+      return [];
+    }
+
+    const tags = [
+      moduleTxt.isTutorial ? "Tutorial" : undefined,
+      moduleTxt.isAsset ? "Asset" : undefined,
+      moduleTxt.isLibrary ? "Library" : undefined,
+      moduleTxt.isSpecial ? "Special" : undefined,
+      moduleTxt.isRendering ? "Rendering" : undefined,
+      moduleTxt.isAugmentation ? "Augmentation" : undefined,
+      moduleTxt.isServerSideOnly ? "Server" : undefined,
+      moduleTxt.isGameplay ? "Gameplay" : undefined,
+    ].filter((x) => x);
+
+    const { id, version, displayName, description, dependencies } = moduleTxt;
+
+    const cover = repo.usesCustomOpenGraphImage
+      ? repo.openGraphImageUrl
+      : undefined;
+
+    const moduleInfo = {
+      id,
+      version,
+      displayName,
+      description,
+      dependencies,
+      tags,
+    };
+
+    moduleInfo.dependencies = moduleInfo.dependencies.map((dep) => ({
+      ...dep,
+      optional: /true/i.test(dep.optional),
+    }));
+
+    const node = {
+      ...repo,
+      usesCustomOpenGraphImage: undefined,
+      openGraphImageUrl: undefined,
+      cover,
+      moduleTxt: moduleInfo,
+      id: createNodeId(`TerasologyModule-${repo.id}`),
+      parent: "__SOURCE__",
+      children: [],
+      internal: {
+        type: "TerasologyModule",
+      },
+    };
+    node.internal.contentDigest = createContentDigest(node);
+
+    let fileNode;
+    if (node.cover) {
       try {
-        moduleTxt = JSON.parse(repo.moduleTxt?.text);
+        fileNode = await createRemoteFileNode({
+          url: node.cover,
+          parentNodeId: node.id,
+          cache,
+          createNodeId,
+          createNode,
+          ext: ".png",
+        });
       } catch (err) {
-        console.warn(
-          `[${PLUGIN_NAME}] Could not parse 'module.txt' of ${repo.url}.`
+        reporter.error(err);
+      }
+      if (fileNode) {
+        node.coverImage___NODE = fileNode.id;
+        reporter.info(
+          `Created remote file node for ${node.name}: ${node.cover}`
         );
-        return;
       }
+    }
+
+    createNode(node);
+
+    // TODO: this is causing some issues I don't understand...
```

Comment:
what kind of issues?

Reply:
```
 ERROR #11321  API.NODE.EXECUTION

"source-terasology-modules" threw an error while running the sourceNodes lifecycle:

Cannot read properties of undefined (reading 'children')



  TypeError: Cannot read properties of undefined (reading 'children')
  
  - public.js:1043 Object.actions.createParentChildLink
    [ModuleSite]/[gatsby]/src/redux/actions/public.js:1043:15
  
  - call-deferred-api.ts:17 callRealApi
    [ModuleSite]/[gatsby]/src/utils/call-deferred-api.ts:17:20
  
  - actions.ts:31 callApi
    [ModuleSite]/[gatsby]/src/state-machines/develop/actions.ts:31:6
  
  - interpreter.js:171 Interpreter._exec
    [ModuleSite]/[xstate]/lib/interpreter.js:171:18
  
  - actions.js:575 handleAction
    [ModuleSite]/[xstate]/lib/actions.js:575:11
  
  - actions.js:603 processBlock
    [ModuleSite]/[xstate]/lib/actions.js:603:24
  
  - Array.map
  
  - actions.js:628 Object.resolveActions
    [ModuleSite]/[xstate]/lib/actions.js:628:52
  
  - StateNode.js:916 StateNode.resolveTransition
    [ModuleSite]/[xstate]/lib/StateNode.js:916:36
  
  - StateNode.js:839 StateNode.transition
    [ModuleSite]/[xstate]/lib/StateNode.js:839:17
  
  - interpreter.js:937 
    [ModuleSite]/[xstate]/lib/interpreter.js:937:28
  
  - serviceScope.js:12 Object.provide
    [ModuleSite]/[xstate]/lib/serviceScope.js:12:16
  
  - interpreter.js:936 Interpreter._nextState
    [ModuleSite]/[xstate]/lib/interpreter.js:936:34
  
  - interpreter.js:108 
    [ModuleSite]/[xstate]/lib/interpreter.js:108:31
  
  - scheduler.js:69 Scheduler.process
    [ModuleSite]/[xstate]/lib/scheduler.js:69:7
  
  - scheduler.js:48 Scheduler.schedule
    [ModuleSite]/[xstate]/lib/scheduler.js:48:10
```

It does not point to a specific line in the plugin, so the error is kind of hidden, which makes it harder to investigate :pensive: I suspect that there is some issue with linking then nodes, and we probably need to `touchNode` to have a placeholder, than create the child nodes and attach them to the touched parent, then create the parent itself, or something like that ...

Reply:
Could it be that this is simply coming from the typo in line 208 leading to the parent being undefined and it cannot find children for taht?

Reply:
Resolved by 39deb981356b3434e30479f8493c610c3095d8b3